src/ErgoDex/Amm/PoolActions.hs

```diff
@@ -0,0 +1,143 @@
+
+import           ErgoDex.Types
+import           ErgoDex.State
+import           ErgoDex.Amm.Orders
+import           ErgoDex.Amm.Pool
+import qualified ErgoDex.Contracts.Pool as P
+import           ErgoDex.Contracts.Types
+import           Cardano.Models
+import           Cardano.Utils
+
+data OrderExecErr = PriceTooHigh deriving (Show)
+
+data PoolActions = PoolActions
+  { runSwap    :: Confirmed Swap    -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
+  , runDeposit :: Confirmed Deposit -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
+  , runRedeem  :: Confirmed Redeem  -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
+  }
+
+mkPoolActions :: PubKeyHash -> PoolActions
+mkPoolActions executorPkh = PoolActions
+  { runSwap    = runSwap' executorPkh
+  , runDeposit = runDeposit' executorPkh
+  , runRedeem  = runRedeem' executorPkh
+  }
+
+runSwap' :: PubKeyHash -> Confirmed Swap -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
+runSwap' executorPkh (Confirmed swapOut Swap{swapExFee=ExFeePerToken{..}, ..}) (Confirmed poolOut pool) = do
+  let
+    poolIn  = FullTxIn poolOut Pay2Script (Just $ Redeemer $ toBuiltinData P.Swap)
+    orderIn = FullTxIn swapOut Pay2Script (Just unitRedeemer)
+    inputs  = [poolIn, orderIn]
+
+    pp@(Predicted nextPoolOut _) = applySwap pool (AssetAmount swapBase swapBaseIn)
+
+    quoteOutput = outputAmount pool (AssetAmount swapBase swapBaseIn)
+
+    executorFee = (assetAmountRawValue quoteOutput) * exFeePerTokenNum `div` exFeePerTokenDen
+    executorOut = TxOutCandidate
+      { txOutCandidateAddress = pubKeyHashAddress executorPkh
+      , txOutCandidateValue   = Ada.lovelaceValueOf executorFee
+      , txOutCandidateDatum   = Nothing
+      }
+
+    rewardOut =
+      TxOutCandidate
+        { txOutCandidateAddress = pubKeyHashAddress swapRewardPkh
+        , txOutCandidateValue   = rewardValue
+        , txOutCandidateDatum   = Nothing
+        }
+      where
+        initValue   = fullTxOutValue swapOut
+        rewardValue = (assetAmountValue quoteOutput) <> (lovelaceSubtract initValue (Ada.Lovelace executorFee))
+
+    outputs = [nextPoolOut, rewardOut, executorOut]
+
+  when (getAmount quoteOutput < swapMinQuoteOut) (Left PriceTooHigh)
+
+  Right (TxCandidate inputs outputs, pp)
+
+runDeposit' :: PubKeyHash -> Confirmed Deposit -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
+runDeposit' executorPkh (Confirmed depositOut Deposit{..}) (Confirmed poolOut pool@Pool{..}) = do
+  let
+    poolIn  = FullTxIn poolOut Pay2Script (Just $ Redeemer $ toBuiltinData P.Deposit)
+    orderIn = FullTxIn depositOut Pay2Script (Just unitRedeemer)
+    inputs  = [poolIn, orderIn]
+
+    (inX, inY) =
+        bimap entryAmount entryAmount $
+          if (assetEntryClass $ fst depositPair) == (unCoin poolCoinX)
+          then depositPair
+          else (snd depositPair, fst depositPair)
```

Comment:
you can use swap from here https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Tuple.html#v%3aswap